### PR TITLE
Logger now creates a log directory if doesn't exist

### PR DESCRIPTION
**Pull request by sauravbanna**
### Description

The logger in ```logger.h``` doesn't create the directory it tries to log to if it doesn't exist. This leads to the error:

```
FILE ERROR:  could not open log file:[/tmp/tbots/yellow_test/thunderbots.log]
		 std::ios_base state = 4
```

In both CI and local tests

### Testing Done

Ran a C++ test and the directory is now created. The error is no longer seen.

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue

### CODE REVIEW

**Comment by Mr-Anyone** [comment]
Finally, this is such a simple fix. LGTM

File: src/software/logger/logger.h [hunk 1]
```diff
@@ -1,5 +1,11 @@
 #pragma once
 
+#if __cplusplus > 201703L
+#include <filesystem>
+#else
+#include <experimental/filesystem>
+#endif
+
```
Comment (itsarune) [inline src/software/logger/logger.h]: I think we compile for c++ 20 through bazel rules so this isn't necessary I think

File: src/software/logger/logger.h [hunk 2]
```diff
@@ -80,6 +86,12 @@ class LoggerSingleton
         // arg. Note: log locations are defaulted to the bazel-out folder due to Bazel's
         // hermetic build principles
 
+        // if log dir doesn't exist, create it
+        if (!std::filesystem::exists(runtime_dir))
+        {
+            std::filesystem::create_directory(runtime_dir);
```
Comment (itsarune) [inline src/software/logger/logger.h]:
```suggestion
            std::filesystem::create_directories(runtime_dir);
```
I hope this fixes your CI issues for this PR

**Review by nimazareian — COMMENTED** [review]
While you're at it, it would be a good idea to update the paths where we store these logs as well. The main one I've noticed is that our GTests are stored here which doesn't make a whole lot of sense

https://github.com/UBC-Thunderbots/Software/blob/5c40f21b067e8f1aaa2794005d43c1aa8bd8444d/src/shared/test_util/tbots_gtest_main.cpp#L14

Maybe instead of `yellow_test` we do something like `gtest_logs`

---

Also, while I was looking around the code for our other runtime directories, I noticed that we have this variable which seems to be unused. If it actually is, could you remove it?

https://github.com/UBC-Thunderbots/Software/blob/5c40f21b067e8f1aaa2794005d43c1aa8bd8444d/src/software/thunderscope/thunderscope_main.py#L327-L331

**Comment by nimazareian** [comment]
@sauravbanna Following up on this PR and your plan for it

**Comment by nimazareian** [comment]
I was just looking at the log directory and noticed that `thunderbots.log` is pretty massive (350+MB for a few minutes of thunderscope), and it turns out that everything we log, including protobufs, are also saved to that file. I think we should wrap `LogRotate::save` in a lambda where it only calls that method if it's one of the default log levels: DEBUG, INFO, WARNING or FATAL

https://github.com/UBC-Thunderbots/Software/blob/7575abc94da6cf88aaa30973334b431bf2c5cc29/src/software/logger/logger.h#L89-L91

File: src/software/logger/logger.h (hunk 1 above)
Comment (sauravbanna) [reply src/software/logger/logger.h]: apparently we need the experimental, it doesnt work otherwise

**Review by nimazareian — DISMISSED** [review]

**Review by nimazareian — DISMISSED** [review]

**Review by williamckha — APPROVED** [review]

**Review by nimazareian — APPROVED** [review]

**Pull request merged**